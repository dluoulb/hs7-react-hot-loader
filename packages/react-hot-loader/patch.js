/* eslint-disable import/no-unresolved */
// export { default } from './lib/patch'
module.exports = require('./lib/patch');
