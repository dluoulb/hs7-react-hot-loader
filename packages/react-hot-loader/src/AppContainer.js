/* eslint-disable global-require, import/no-mutable-exports */

let exportedModule

if (!module.hot || process.env.NODE_ENV === 'production') {
  exportedModule = require('./AppContainer.prod').default
} else {
  exportedModule = require('./AppContainer.dev').default
}

export default exportedModule
