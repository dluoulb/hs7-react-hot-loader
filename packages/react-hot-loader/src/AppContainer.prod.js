/* eslint-disable react/prop-types */

import React, { Component } from 'react'

class AppContainer extends Component {
  render() {
    if (this.props.component) {
      return <this.props.component {...this.props.props} />
    }

    return React.Children.only(this.props.children)
  }
}

export default AppContainer
