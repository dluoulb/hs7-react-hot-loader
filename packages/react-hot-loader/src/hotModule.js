const moduleOpened = {}

const moduleEntry = {
  enter(sourceModule) {
    const moduleId = sourceModule.id
    moduleOpened[moduleId] = true
  },

  leave(sourceModule) {
    delete moduleOpened[sourceModule.id]
  },
}

export const isModuleOpened = module => !!moduleOpened[module.id]

export default moduleEntry
