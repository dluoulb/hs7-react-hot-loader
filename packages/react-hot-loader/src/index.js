/* eslint-disable global-require, import/no-mutable-exports */

let exportedModule

if (!module.hot || process.env.NODE_ENV === 'production') {
  exportedModule = require('./index.prod')
} else {
  exportedModule = require('./index.dev')
}

module.exports = exportedModule;