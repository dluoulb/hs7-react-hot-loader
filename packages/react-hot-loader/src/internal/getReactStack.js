/* eslint-disable no-underscore-dangle */

import { getReactInstance, getReactComponent } from './reactUtils'

function pushState(stack, instance, element) {
  stack.type = instance.type
  stack.tag = instance.tag
  stack.children = []
  stack.instance = element || getReactComponent(instance) || stack
}

// these function might be obsolete
function traverseRenderedChildren(internalInstance, stack) {
  if (internalInstance._currentElement) {
    pushState(
      stack,
      internalInstance._currentElement,
      internalInstance._instance,
    )
  }

  if (internalInstance._renderedComponent) {
    const childStack = {}
    traverseRenderedChildren(internalInstance._renderedComponent, childStack)
    stack.children.push(childStack)
  } else if (internalInstance._renderedChildren) {
    Object.keys(internalInstance._renderedChildren).forEach(key => {
      const childStack = {}
      traverseRenderedChildren(
        internalInstance._renderedChildren[key],
        childStack,
      )
      stack.children.push(childStack)
    })
  }
}

function hydrateStack(instance) {
  const internalInstance = instance._reactInternalInstance
  const stack = {}
  traverseRenderedChildren(internalInstance, stack)
  return stack
}

function traverseTree(root, stack) {
  pushState(stack, root)
  const node = root
  if (node.child) {
    let { child } = node
    do {
      const childStack = {}
      traverseTree(child, childStack)
      stack.children.push(childStack)
      child = child.sibling
    } while (child)
  }
}

// modern react tree
function hydrateTree(root) {
  const stack = {}
  traverseTree(root, stack)
  return stack
}

function getReactStack(instance) {
  const root = getReactInstance(instance)
  if (typeof root.tag !== 'number') {
    // Traverse stack-based React tree.
    return hydrateStack(instance)
  }
  return hydrateTree(root)
}

export default getReactStack
