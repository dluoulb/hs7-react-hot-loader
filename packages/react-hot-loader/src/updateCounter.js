let generation = 0 ;

export const didUpdate = () => generation++;
export const getGeneration = () => generation;