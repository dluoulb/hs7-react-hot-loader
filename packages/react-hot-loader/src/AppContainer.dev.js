import React, { Component } from 'react'
import PropTypes from 'prop-types'
import deepForceUpdate from 'react-deep-force-update'
import {getGeneration} from "./updateCounter";

class AppContainer extends Component {
  constructor(props) {
    super(props)

    if (
      props.warnings === false &&
      typeof __REACT_HOT_LOADER__ !== 'undefined'
    ) {
      __REACT_HOT_LOADER__.warnings = props.warnings
    }

    this.state = { error: null }
  }

  componentDidMount() {
    if (typeof __REACT_HOT_LOADER__ === 'undefined') {
      console.error(
        'React Hot Loader: It appears that "react-hot-loader/patch" ' +
          'did not run immediately before the app started. Make sure that it ' +
          'runs before any other code. For example, if you use Webpack, ' +
          'you can add "react-hot-loader/patch" as the very first item to the ' +
          '"entry" array in its config. Alternatively, you can add ' +
          'require("react-hot-loader/patch") as the very first line ' +
          'in the application code, before any other imports.',
      )
    }
  }

  componentWillReceiveProps() {
    if (this.state.generation!==getGeneration()) {
      // Hot reload is happening.
      // Retry rendering!
      this.setState({
        error: null,
        generation: getGeneration()
      })
      // Force-update the whole tree, including
      // components that refuse to update.
      deepForceUpdate(this)
    }
  }

  shouldComponentUpdate(prevProps, prevState) {
    // Don't update the component if the state had an error and still has one.
    // This allows to break an infinite loop of error -> render -> error -> render
    // https://github.com/gaearon/react-hot-loader/issues/696
    if (prevState.error && this.state.error) {
      return false
    }

    return true
  }

  // This hook is going to become official in React 15.x.
  // In 15.0, it only catches errors on initial mount.
  // Later it will work for updates as well:
  // https://github.com/facebook/react/pull/6020
  /* eslint-disable camelcase */
  unstable_handleError(error) {
    this.componentDidCatch(error)
  }
  /* eslint-enable camelcase */

  componentDidCatch(error) {
    this.setState({
      error,
    })
  }

  render() {
    const { error } = this.state

    if (this.props.errorReporter && error) {
      console.error(error)
      return <this.props.errorReporter error={error} />
    } else if (error) {
      console.error(error)
    }

    return React.Children.only(this.props.children)
  }
}

AppContainer.propTypes = {
  children(props) {
    if (React.Children.count(props.children) !== 1) {
      return new Error(
        'Invalid prop "children" supplied to AppContainer. ' +
          'Expected a single React element with your appâ€™s root component, e.g. <App />.',
      )
    }

    return undefined
  },
  errorReporter: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  warnings: PropTypes.bool,
}

export default AppContainer
