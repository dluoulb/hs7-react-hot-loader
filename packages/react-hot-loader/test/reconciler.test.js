/* eslint-env jest */

import React, { Component } from 'react'
import { mount } from 'enzyme'
import '../src/patch.dev'
import AppContainer from '../src/AppContainer.dev'
import { didUpdate } from '../src/updateCounter'
import getReactStack from '../src/internal/getReactStack'
import { areComponentsEqual } from '../src/utils.dev'
import RHL from '../src/reactHotLoader'

const spyComponent = (render, displayName, key) => {
  const mounted = jest.fn()
  const unmounted = jest.fn()
  const willUpdate = jest.fn()

  class TestingComponent extends Component {
    componentWillMount() {
      mounted()
    }

    componentWillUpdate(nextProps, nextState) {
      willUpdate(nextProps, nextState, this.props, this.state)
    }

    componentWillUnmount() {
      unmounted()
    }

    /* eslint-disable */
    __reactstandin__regenerateByEval(key, code) {
      this[key] = eval(code)
    }
    /* eslint-enable */

    render() {
      return render(this.props)
    }
  }

  if (displayName) {
    TestingComponent.displayName = displayName
  }

  return {
    Component: TestingComponent,
    willUpdate,
    mounted,
    unmounted,
    key,
  }
}

describe('reconciler', () => {
  describe('Application', () => {
    it('should regenerate internal component', () => {
      const root = spyComponent(
        ({ children }) => <div>{children}</div>,
        'root',
        'root',
      )

      const first = spyComponent(
        ({ children }) => <b>{children}</b>,
        'test',
        '1',
      )
      const second = spyComponent(() => <u>REPLACED</u>, 'test', '2')
      const third = spyComponent(() => <u>NEW ONE</u>, 'somethingElse', '3')

      let currentComponent = first
      const currentProps = {}

      const ComponentSwap = props => {
        const { Component } = currentComponent
        return (
          <blockquote>
            <Component {...props} {...currentProps} />
          </blockquote>
        )
      }

      const App = () => (
        <root.Component>
          <ComponentSwap>42</ComponentSwap>
        </root.Component>
      )

      const wrapper = mount(
        <AppContainer>
          <App />
        </AppContainer>,
      )

      // mount and perform first checks
      expect(wrapper.find(<first.Component />.type).length).toBe(1)
      expect(root.mounted).toHaveBeenCalledTimes(1)
      expect(first.mounted).toHaveBeenCalledTimes(1)

      // replace with `the same` component
      currentComponent = second
      // they are different
      expect(areComponentsEqual(first.Component, second.Component)).toBe(false)

      currentProps.newProp = true
      didUpdate()
      wrapper.setProps({ update: 'now' })
      // not react-stand-in merge them together
      expect(areComponentsEqual(first.Component, second.Component)).toBe(true)
      expect(wrapper.find(<first.Component />.type).length).toBe(1)
      expect(wrapper.find(<second.Component />.type).length).toBe(1)

      expect(root.mounted).toHaveBeenCalledTimes(1)
      expect(first.unmounted).toHaveBeenCalledTimes(0)
      expect(second.mounted).toHaveBeenCalledTimes(0)
      expect(second.willUpdate).toHaveBeenCalledTimes(2)
      expect(second.willUpdate.mock.calls[0]).toEqual([
        { children: '42' },
        null,
        { children: '42' },
        null,
      ])
      expect(second.willUpdate.mock.calls[1]).toEqual([
        { children: '42', newProp: true },
        null,
        { children: '42' },
        null,
      ])

      // replace with a different component
      currentComponent = third
      didUpdate()
      wrapper.setProps({ update: 'now' })
      expect(wrapper.find(<third.Component />.type).length).toBe(1)
      // first will never be unmounted
      expect(first.unmounted).toHaveBeenCalledTimes(0)
      expect(second.unmounted).toHaveBeenCalledTimes(1)
      expect(third.mounted).toHaveBeenCalledTimes(1)

      expect(areComponentsEqual(first.Component, third.Component)).toBe(false)
      expect(areComponentsEqual(second.Component, third.Component)).toBe(false)
    })

    it('should regenerate internal component', () => {
      const first = spyComponent(
        ({ children }) => <b>{children}</b>,
        'test',
        '1',
      )

      const second = spyComponent(() => <u>REPLACED</u>, 'test', '2')

      let currentComponent = first

      const ComponentSwap = props => {
        const { Component } = currentComponent
        return (
          <blockquote>
            <Component {...props} />
          </blockquote>
        )
      }

      const App = () => (
        <div>
          <h1>working</h1>
          <ComponentSwap>
            <ComponentSwap>42</ComponentSwap>
          </ComponentSwap>
        </div>
      )

      const wrapper = mount(
        <AppContainer>
          <App />
        </AppContainer>,
      )

      currentComponent = second
      didUpdate()
      wrapper.setProps({ update: 'now' })

      expect(first.unmounted).toHaveBeenCalledTimes(0)
      expect(second.mounted).toHaveBeenCalledTimes(0)
    })

    it('should handle function as a child', () => {
      const first = spyComponent(
        ({ children }) => <b>{children(0)}</b>,
        'test',
        '1',
      )
      const second = spyComponent(
        ({ children }) => <u>{children(1)}</u>,
        'test',
        '2',
      )

      let currentComponent = first
      const ComponentSwap = props => {
        const { Component } = currentComponent
        return (
          <blockquote>
            <Component {...props} />
          </blockquote>
        )
      }

      const App = () => (
        <div>
          <h1>working</h1>
          <ComponentSwap>{x => 42 + x}</ComponentSwap>
        </div>
      )

      const wrapper = mount(
        <AppContainer>
          <App />
        </AppContainer>,
      )

      expect(wrapper.text()).toContain(42)

      currentComponent = second
      didUpdate()
      wrapper.setProps({ update: 'now' })

      expect(first.unmounted).toHaveBeenCalledTimes(0)
      expect(second.mounted).toHaveBeenCalledTimes(0)
      expect(wrapper.text()).toContain(43)
    })
  })

  describe('hydrate', () => {
    it('should hydrate', () => {
      const Transform = ({ children }) => <section>42 + {children}</section>
      const One = ({ children }) => <section>1 == {children(1)}</section>

      RHL.disableProxyCreation = true
      const wrapper = mount(
        <AppContainer>
          <div>
            <div>
              <Transform>
                <div>
                  <div>42</div>
                </div>
              </Transform>
            </div>
            <div>
              <One>{one => one}</One>
            </div>
            <div>
              the lazy brown fox
              <p>jumped over the hedge</p>
            </div>
          </div>
        </AppContainer>,
      )
      RHL.disableProxyCreation = false
      const { instance, children } = getReactStack(wrapper.instance())
      expect(children).toMatchSnapshot()
      expect(instance).not.toBe(null)
    })
  })
})
