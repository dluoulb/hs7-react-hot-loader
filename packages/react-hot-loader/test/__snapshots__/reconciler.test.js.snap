// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "instance": [Circular],
                    "tag": 6,
                    "type": null,
                  },
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "instance": [Circular],
                        "tag": 5,
                        "type": "div",
                      },
                    ],
                    "instance": [Circular],
                    "tag": 5,
                    "type": "div",
                  },
                ],
                "instance": [Circular],
                "tag": 5,
                "type": "section",
              },
            ],
            "instance": [Circular],
            "tag": 1,
            "type": [Function],
          },
        ],
        "instance": [Circular],
        "tag": 5,
        "type": "div",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "instance": [Circular],
                    "tag": 6,
                    "type": null,
                  },
                  Object {
                    "children": Array [],
                    "instance": [Circular],
                    "tag": 6,
                    "type": null,
                  },
                ],
                "instance": [Circular],
                "tag": 5,
                "type": "section",
              },
            ],
            "instance": [Circular],
            "tag": 1,
            "type": [Function],
          },
        ],
        "instance": [Circular],
        "tag": 5,
        "type": "div",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "instance": [Circular],
            "tag": 6,
            "type": null,
          },
          Object {
            "children": Array [],
            "instance": [Circular],
            "tag": 5,
            "type": "p",
          },
        ],
        "instance": [Circular],
        "tag": 5,
        "type": "div",
      },
    ],
    "instance": [Circular],
    "tag": 5,
    "type": "div",
  },
]
`;
